File: C++/Day19.cpp
#ifdef DAY19

#include "AH.h"

namespace Day19
{
	
	struct Instruction
	{
		char mode;
		char comp;
		int value;
		std::string out;

		Instruction(const char mode, const char comp, const int value,
		const std::string out) :
		  mode(mode), comp(comp), value(value), out(out) {};
		Instruction() : comp('#'), value(-1), out("BAD") {};
	};

	std::map<char, int> modeToIndex { {'x', 0}, {'m', 1}, {'a', 2}, {'s', 3} };

	struct Workflow
	{
		std::vector<Instruction> ins;
		std::string fallback;
	};

	typedef std::vector<std::pair<int, int>> Ranges;

	typedef std::vector<int> Part;

	
	std::map<std::string, Workflow> getWorkflows(const std::string ss)
	{
		std::map<std::string, Workflow> wfs;
		
		const auto ll = AH::SplitOnString(ss, "|");
		for (auto l : ll) {
			const auto bra = l.find('{');
			const auto ket = l.find('}');
			const auto prefix = l.substr(0, bra);
			const auto contents = l.substr(bra + 1, ket - bra - 1);

			const auto cs = AH::SplitOnString(contents, ",");

			Workflow wf;
			for (auto c : cs) {
				if (c.find(':') != std::string::npos) {
					const auto qq = AH::SplitOnString(c, ":");
					const char mode = qq[0].at(0);
					const char comp = qq[0].at(1);
					const int value = stoi(qq[0].substr(2));
					const std::string out = qq[1];
					const Instruction in(mode, comp, value, out);

					wf.ins.push_back(in);
				} else {
					wf.fallback = c;
				}
			}
			wfs[prefix] = wf;
		}

		return wfs;
	}

	std::vector<Part> getParts(const std::string ss)
	{
		const auto ll = AH::SplitOnString(ss, "|");
		std::vector<Part> parts;

		for (auto l : ll) {
			Part pt(4, -1);
			const std::string trimmed = l.substr(1, l.size() - 2);
			const auto qq = AH::SplitOnString(trimmed, ",");

			for (auto q : qq) {
				const auto mode = q.at(0);
				const auto int_str = q.substr(2);
				const int i = stoi(int_str);
				pt[modeToIndex[mode]] = i;
			}

			parts.push_back(pt);
		}

		return parts;
	}

	std::string applySingleWorkflow(const Part p, const Workflow wf)
	{
		for (auto is : wf.ins) { // resolve instructions in order
			const auto idx = modeToIndex[is.mode];
			const auto value = p[idx];

			const auto c = is.comp;
			if (c == '>') {
				if (value > is.value) {
					return is.out;
				}
			} else {
				if (value < is.value) {
					return is.out;
				}
			}
		}
		// failed to find matching order so use fallback
		return wf.fallback;
	}

	int applyWorkflows(const Part p, const std::map<std::string, Workflow> wfs)
	{
		Workflow wf = wfs.at("in"); // we start here
		std::string here = applySingleWorkflow(p, wf);

		while ((here != "A") && (here != "R")) {
			wf = wfs.at(here);
			here = applySingleWorkflow(p, wf);
		}

		if (here == "A") {
			return std::accumulate(std::begin(p), std::end(p), 0);
		} else {
			return 0;
		}
	}

	int64_t rangeSize(const Ranges r)
	{
		int64_t ret = 1;
		for (auto [lo, hi] : r) {
			ret *= (hi - lo + 1);
		}

		return ret;
	}

	int64_t applyWorkToRanges(Ranges rng, const std::map<std::string, Workflow> wfs,
		const std::string here)
	{
		int64_t ret = 0;
		Workflow wf = wfs.at(here); // we start here
		for (auto i : wf.ins) {
			const auto c = i.comp;
			if (c != '#') { // this is the final instruction, with no conditions
				if (c == '>') {
					auto new_rng = rng;
					auto idx = modeToIndex[i.mode];
					if (new_rng[idx].second > i.value) {
						new_rng[idx].first = std::max(new_rng[idx].first, i.value + 1);
						if (i.out == "A") {
							ret += rangeSize(new_rng);
						} else if (i.out != "R") {
							ret += applyWorkToRanges(new_rng, wfs, i.out);
						}
						rng[idx].second = std::min(rng[idx].second, i.value);
					}
				}
				if (c == '<') {
					auto new_rng = rng;
					auto idx = modeToIndex[i.mode];
					if (new_rng[idx].first < i.value) {
						new_rng[idx].second = std::min(new_rng[idx].second, i.value - 1);
						if (i.out == "A") {
							ret += rangeSize(new_rng);
						} else if (i.out != "R") {
							ret += applyWorkToRanges(new_rng, wfs, i.out);
						}
						rng[idx].first = std::max(rng[idx].first, i.value);
					}
				}
			} else{
			}
		}
		// applied all the legit instruction so finally apply the fallback condition
		if (wf.fallback == "A") {
			ret += rangeSize(rng);
		} else if (wf.fallback != "R") {
			ret += applyWorkToRanges(rng, wfs, wf.fallback);
		}
		return ret;
	}

	int Run(const std::string& filename)
	{
		const std::vector<std::string> ls = AH::ReadTextFile(filename);
		const auto gs = AH::ParseLineGroups(ls, '|');
		const auto wfs = getWorkflows(gs[0]);
		const auto parts = getParts(gs[1]);

		int64_t part1 = 0;
		int64_t part2 = 0;

		for (auto p : parts) {
			part1 += applyWorkflows(p, wfs);
		}

		Ranges rng = { {1,4000}, {1,4000}, {1,4000}, {1,4000} };
		part2 = applyWorkToRanges(rng, wfs, "in");

		AH::PrintSoln(19, part1, part2);

		return 0;
	}

}
   
#endif
 

File: Go/Day19/Day19.go
//go:build d19
// +build d19

package Day19

import (
	"strconv"
	"strings"

	AH "AoC/adventhelper"
)

type mode int
type comp int

const (
	x mode = iota
	m
	a
	s
	f
)

func runeToMode(r rune) mode {
	switch r {
	case 'x':
		return x
	case 'm':
		return m
	case 'a':
		return a
	case 's':
		return s
	}

	return f
}

const (
	lt comp = iota
	gt
	none
)

func runeToComp(r rune) comp {
	switch r {
	case '<':
		return lt
	case '>':
		return gt
	}

	return none
}

type Instruction struct {
	m mode
	c comp
	v int
	o string
}

type WorkFlow struct {
	ins      []Instruction
	fallback string
}

type Range struct {
	lo, hi int
}

func getWorkflows(ss string) map[string]WorkFlow {
	wfs := make(map[string]WorkFlow)
	ls := strings.Split(ss, "|")
	for _, l := range ls {
		bra := strings.Index(l, "{")
		prefix := l[:bra]
		contents := l[bra+1 : len(l)-1]
		cs := strings.Split(contents, ",")
		wf := WorkFlow{ins: []Instruction{}, fallback: "X"}
		for _, cc := range cs {
			if strings.Index(cc, ":") > 0 {
				qq := strings.Split(cc, ":")
				m := runeToMode(AH.RuneAt(qq[0], 0))
				c := runeToComp(AH.RuneAt(qq[0], 1))
				v, _ := strconv.Atoi(qq[0][2:])
				o := qq[1]

				ins := Instruction{m, c, v, o}
				wf.ins = append(wf.ins, ins)
			} else {
				wf.fallback = cc
			}
		}

		wfs[prefix] = wf
	}

	return wfs
}

func getParts(ss string) [][]int {
	ls := strings.Split(ss, "|")
	parts := [][]int{}

	for _, l := range ls {
		pt := make([]int, 4)
		trimmed := l[1 : len(l)-1]
		qq := strings.Split(trimmed, ",")

		for i, q := range qq {
			val, _ := strconv.Atoi(q[2:])
			pt[i] = val
		}

		parts = append(parts, pt)
	}

	return parts
}

func applyWorkflows(p []int, wfs map[string]WorkFlow, here string) int {
	for (here != "A") && (here != "R") {
		wf := wfs[here]

		here = wf.fallback
		for _, is := range wf.ins {
			v := p[is.m]
			if ((is.c == gt) && (v > is.v)) || ((is.c == lt) && (v < is.v)) {
				here = is.o
				break
			}
		}
	}

	if here == "A" {
		sum := 0
		for _, pp := range p {
			sum += pp
		}
		return sum
	}

	return 0
}

func rangeSize(rs []Range) int {
	ret := 1
	for _, r := range rs {
		ret *= (r.hi - r.lo + 1)
	}

	return ret
}

func applyWorkflowToRanges(rngs []Range, wfs map[string]WorkFlow, here string) int {
	ret := 0
	wf := wfs[here]
	for _, i := range wf.ins {
		newRngs := make([]Range, 4)
		copy(newRngs, rngs)
		switch i.c {
		case gt:
			if newRngs[i.m].hi > i.v {
				newRngs[i.m].lo = AH.Max(newRngs[i.m].lo, i.v+1)
				if i.o == "A" {
					ret += rangeSize(newRngs)
				} else if i.o != "R" {
					ret += applyWorkflowToRanges(newRngs, wfs, i.o)
				}
				rngs[i.m].hi = AH.Min(rngs[i.m].hi, i.v)
			}
		case lt:
			if newRngs[i.m].lo < i.v {
				newRngs[i.m].hi = AH.Min(newRngs[i.m].hi, i.v-1)
				if i.o == "A" {
					ret += rangeSize(newRngs)
				} else if i.o != "R" {
					ret += applyWorkflowToRanges(newRngs, wfs, i.o)
				}
				rngs[i.m].lo = AH.Max(rngs[i.m].lo, i.v)
			}
		}
	}
	if wf.fallback == "A" {
		ret += rangeSize(rngs)
	} else if wf.fallback != "R" {
		ret += applyWorkflowToRanges(rngs, wfs, wf.fallback)
	}

	return ret
}

func Run() {
	gs, _ := AH.ParseLineGroups("../input/input19.txt", "|")
	wfs := getWorkflows(gs[0])
	parts := getParts(gs[1])

	part1 := 0
	for _, p := range parts {
		part1 += applyWorkflows(p, wfs, "in")
	}
	rngs := []Range{{1, 4000}, {1, 4000}, {1, 4000}, {1, 4000}}

	AH.PrintSoln(19, part1, applyWorkflowToRanges(rngs, wfs, "in"))

	return
}
